File: app/scripts/manager.py
from typing import Optional
from PySide6.QtCore import QObject, Signal
from app.models.project import Project
from app.process.runner import ProcessRunner


class ScriptManager(QObject):
    script_started = Signal(str)  # script_name
    script_finished = Signal(str, int)  # script_name, exit_code
    log_output = Signal(str, bool)  # message, is_error

    def __init__(self, project: Project, parent=None):
        super().__init__(parent)
        self._project = project
        self._runner = ProcessRunner(self)
        self._active_name: Optional[str] = None
        self._active_command: Optional[str] = None

        self._runner.process_started.connect(self._on_started)
        self._runner.output_ready.connect(lambda msg: self.log_output.emit(f"[{self._active_name}] {msg}", False))
        self._runner.error_ready.connect(lambda msg: self.log_output.emit(f"[{self._active_name}] {msg}", True))
        self._runner.process_finished.connect(self._on_finished)
        self._runner.process_error.connect(lambda msg: self.log_output.emit(f"[{self._active_name}] ERROR: {msg}", True))

    @property
    def project(self) -> Project:
        return self._project

    def update_project(self, project: Project):
        self._project = project

    def is_running(self) -> bool:
        return self._runner.is_running()

    @property
    def active_script_name(self) -> Optional[str]:
        return self._active_name if self.is_running() else None

    def run_script(self, name: str, command: str):
        if self.is_running():
            self.log_output.emit(f"⚠️ Cannot start '{name}': script '{self._active_name}' is already running.", True)
            return

        self._active_name = name
        self._active_command = command
        self.log_output.emit(f"⚡ Starting script '{name}': {command}", False)
        self._runner.start(command, self._project.path)

    def stop(self):
        if self.is_running():
            name = self._active_name or "Script"
            self.log_output.emit(f"🛑 Stopping script '{name}'...", False)
            self._runner.stop()

    def _on_started(self):
        if self._active_name:
            self.script_started.emit(self._active_name)

    def _on_finished(self, exit_code: int, status_desc: str):
        finished_name = self._active_name or "Script"
        msg = f"✓ Script '{finished_name}' finished (code {exit_code})" if exit_code == 0 else f"✕ Script '{finished_name}' exited with code {exit_code} ({status_desc})"
        self.log_output.emit(msg, exit_code != 0)
        self.script_finished.emit(finished_name, exit_code)
        self._active_name = None
        self._active_command = None

File: devmanager-app/internal/scripts/manager.go
// Package scripts porta app/scripts/manager.py: un proceso de script por
// proyecto, con prefijo [nombre] en el log y stop con intenci├│n expl├¡cita.
package scripts

import (
	"strconv"
	"sync"

	"github.com/d-l-n/devmanager/internal/models"
	"github.com/d-l-n/devmanager/internal/process"
)

type Callbacks struct {
	OnScriptStarted  func(name string)
	OnScriptFinished func(name string, exitCode int)
	OnLog            func(msg string, isError bool)
}

type Manager struct {
	mu sync.Mutex

	project models.Project
	cb      Callbacks

	runner        *process.Runner
	activeName    string
	activeCommand string
	stopRequested bool
}

func NewManager(project models.Project, cb Callbacks) *Manager {
	m := &Manager{project: project, cb: cb}
	m.runner = process.NewRunner(process.RunnerCallbacks{
		OnStarted:  m.onRunnerStarted,
		OnStdout:   func(line string) { m.emitPrefixed(line, false) },
		OnStderr:   func(line string) { m.emitPrefixed(line, true) },
		OnFinished: m.onRunnerFinished,
		OnError:    func(desc string) { m.emitError(desc) },
	})
	return m
}

func (m *Manager) UpdateProject(p models.Project) {
	m.mu.Lock()
	defer m.mu.Unlock()
	m.project = p
}

func (m *Manager) IsRunning() bool { return m.runner.IsRunning() }

// ActiveScriptName replica active_script_name: vac├¡o cuando no corre.
func (m *Manager) ActiveScriptName() string {
	if !m.IsRunning() {
		return ""
	}
	m.mu.Lock()
	defer m.mu.Unlock()
	return m.activeName
}

// RunScript replica run_script: un solo script activo por proyecto.
func (m *Manager) RunScript(name, command string) {
	if m.IsRunning() {
		m.mu.Lock()
		active := m.activeName
		m.mu.Unlock()
		m.log("ÔÜá´©Å Cannot start '"+name+"': script '"+active+"' is already running.", true)
		return
	}

	m.mu.Lock()
	m.activeName = name
	m.activeCommand = command
	m.stopRequested = false
	path := m.project.Path
	m.mu.Unlock()

	m.log("ÔÜí Starting script '"+name+"': "+command, false)
	_ = m.runner.Start(command, path, nil)
}

// Stop replica stop(): marca intenci├│n ANTES de matar (taskkill => CrashExit
// no cuenta como fallo del script).
func (m *Manager) Stop() {
	if !m.IsRunning() {
		return
	}
	m.mu.Lock()
	name := m.activeName
	if name == "" {
		name = "Script"
	}
	m.stopRequested = true
	m.mu.Unlock()

	m.log("­ƒøæ Stopping script '"+name+"'...", false)
	m.runner.Stop()
}

func (m *Manager) log(msg string, isError bool) {
	if m.cb.OnLog != nil {
		m.cb.OnLog(msg, isError)
	}
}

// emitPrefixed replica los lambdas output_ready/error_ready: "[name] msg".
func (m *Manager) emitPrefixed(msg string, isError bool) {
	name := m.activeNameForLogs()
	m.log("["+name+"] "+msg, isError)
}

// emitError replica el lambda process_error: "[name] ERROR: msg".
func (m *Manager) emitError(desc string) {
	name := m.activeNameForLogs()
	m.log("["+name+"] ERROR: "+desc, true)
}

func (m *Manager) activeNameForLogs() string {
	m.mu.Lock()
	defer m.mu.Unlock()
	return m.activeName
}

func (m *Manager) onRunnerStarted() {
	m.mu.Lock()
	name := m.activeName
	m.mu.Unlock()
	if name != "" && m.cb.OnScriptStarted != nil {
		m.cb.OnScriptStarted(name)
	}
}

// onRunnerFinished replica _on_finished.
//
// DIVERGENCIA INTENCIONAL vs Python (#8): en Python _on_finished no distingue
// stop; el runner moribundo (taskkill => exit no-0) pasa el c├│digo REAL a
// script_finished, lanzando notificaci├│n de fallo aunque el usuario cancelara.
// Aqu├¡, si hubo Stop(), reportamos script_finished(name, 0) ÔåÆ cierre limpio sin
// notificaci├│n de fallo cosm├®tica. Decisi├│n expl├¡cita de producto.
func (m *Manager) onRunnerFinished(exitCode int, statusDesc string) {
	m.mu.Lock()
	finishedName := m.activeName
	if finishedName == "" {
		finishedName = "Script"
	}
	wasStopped := m.stopRequested
	m.mu.Unlock()

	if wasStopped {
		m.mu.Lock()
		m.stopRequested = false
		m.activeName = ""
		m.activeCommand = ""
		m.mu.Unlock()
		m.log("­ƒøæ Script '"+finishedName+"' stopped by user.", false)
		if m.cb.OnScriptFinished != nil {
			m.cb.OnScriptFinished(finishedName, 0)
		}
		return
	}

	msg := ""
	isErr := exitCode != 0
	if exitCode == 0 {
		msg = "Ô£ô Script '" + finishedName + "' finished (code 0)"
	} else {
		msg = "Ô£ò Script '" + finishedName + "' exited with code " + strconv.Itoa(exitCode) + " (" + statusDesc + ")"
	}
	m.log(msg, isErr)

	if m.cb.OnScriptFinished != nil {
		m.cb.OnScriptFinished(finishedName, exitCode)
	}
	m.mu.Lock()
	m.activeName = ""
	m.activeCommand = ""
	m.mu.Unlock()
}
